### ISSUE TITLE
Wrong clustering columns ordering during table create, user input not respected

### ISSUE THREAD

**Issue by rtatavarty** [outworkers/phantom#625]
Looks like I am hitting the same issue #618 in version 2.1.2. I am creating a new issue for this instead of working on a closed one.

The table has a few clustering keys with their own clustering order as show below:

```scala
case class UsageByDev(
  key: String,
  dh: String,
  date: DateTime,
  aid: String,
  did: String,
  st: DateTime,
  et: DateTime)

class UsageByDevs extends CassandraTable[IUsageByDevs, UsageByDev] {
  object key extends StringColumn(this) with PartitionKey
  object dh extends StringColumn(this) with PartitionKey
  object date extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Descending
  object aid extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending
  object did extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending
  object st extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Ascending
  object et extends DateTimeColumn(this)
}
```

But ```table.create.ifNotExists()``` threw errors as follows:

```log
[error] c.o.phantom - The order of columns in the CLUSTERING ORDER directive must be the one of the clustering key (date must appear before aid)
```

Upon turning on tracing in Cassandra and capturing the queries, the following was tried during create:

```sql
CREATE TABLE IF NOT EXISTS harvest_production.ns_app_usage_by_devices (aid text, dh text, st timestamp, key text, et timestamp, date timestamp, did text,
PRIMARY KEY ((key, dh), date, aid, did, st))
WITH CLUSTERING ORDER BY (st ASC, date DESC, did ASC, aid ASC);
```
This clearly shows that the ordering is changed and Cassandra is throwing the appropriate error.

I have got a few tables like this and it fails on all of them.

**Comment by melodious** [comment]
I am having the same issue. Using version 2.1.1.

**Comment by melodious** [comment]
Is there any version where this does work? I've tried 2.1.0, 2.1.1 and 2.1.2 and none of them work.

**Comment by rtatavarty** [comment]
I was working in 1.22, guess that's too old now. For now I just settled with ```PRIMARY KEY``` and dropped the ```CLUSTERING ORDER ASCENDING``` and it working without any problem.

**Comment by alexflav23** [comment]
@rtatavarty @melodious Fixed in 2.1.3, on Maven Central in 1 hour.

**Comment by melodious** [comment]
This is still not working for me. I have this table definition: 
```scala
case class FoobarTable(targetSite: String,
                                       timePartition: String,
                                       lastTimestamp: DateTime,
                                       firstTimestamp: DateTime,
                                       testType: String,
                                       testSubtype: String,
                                       name: String,
                                       target: String,
                                       testSpecArgs: String,
                                       json: String)

class FoobarTables extends CassandraTable[AbstractFoobarTables, FoobarTable] {

  object targetSite extends StringColumn(this) with PartitionKey

  object timePartition extends StringColumn(this) with PartitionKey

  object lastTimestamp extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Descending

  object firstTimestamp extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Descending

  object testType extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object testSubtype extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object name extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object target extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object testSpecArgs extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object json extends StringColumn(this)

  override def fromRow(row: Row): FoobarTable = {
    FoobarTable(
      targetSite = targetSite(row),
      timePartition = timePartition(row),
      lastTimestamp = lastTimestamp(row),
      firstTimestamp = firstTimestamp(row),
      testType = testType(row),
      testSubtype = testSubtype(row),
      name = name(row),
      target = target(row),
      testSpecArgs = testSpecArgs(row),
      json = json(row)
    )
  }
}
```

I am getting this error:
```
The order of columns in the CLUSTERING ORDER directive must be the one of the clustering key (name must appear before lasttimestamp)
```
And the SQL produced is this:
```
CREATE TABLE somekeyspace.FoobarTable (targetSite text, timePartition text, lastTimestamp timestamp, firstTimestamp timestamp, testType text, testSubtype text, name text, target text, testSpecArgs text, json text, PRIMARY KEY ((targetSite, timePartition), name, testSpecArgs, testSubtype, testType, lastTimestamp, target, firstTimestamp)) WITH CLUSTERING ORDER BY (lastTimestamp DESC, firstTimestamp DESC, testType ASC, testSubtype ASC, name ASC, target ASC, testSpecArgs ASC);
```

**Comment by alexflav23** [comment]
@melodious I'm not sure what is not working because phantom is respecting the order of keys you are typing in, that's what it's meant to do, and it looks like it's doing just that.

**Comment by melodious** [comment]
I'm assuming the issue is with the order of the primary keys. They don't match the order of the clustering order. I don't have a local cassandra set up just yet to prove that but from looking at the CQL it seems like the one thing that is not in order that could trigger this error.

**Comment by alexflav23** [comment]
Hi @melodious First of all `fromRow` should be automatically derived in 2.0.x+, so you don't need to manually define it. Next up phantom will simply respect your write order, so the cluster key will just look exactly like what you write in the DSL code. The macro simply works to respect your write order.

If that has culprits or yield invalid Cassandra queries, that would be interesting to understand. If you are implying there is a mismatch between the primary keys and the clustering keys, that would be a problem indeed, but the same inference mechanism is used.

As a last notion, you actually don't need to define both clustering and primary on the same column. So your table should look like:

```scala
class FoobarTables extends CassandraTable[AbstractFoobarTables, FoobarTable] {

  object targetSite extends StringColumn(this) with PartitionKey

  object timePartition extends StringColumn(this) with PartitionKey

  object lastTimestamp extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Descending

  object firstTimestamp extends DateTimeColumn(this) with PrimaryKey with ClusteringOrder with Descending

  object testType extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object testSubtype extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object name extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object target extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object testSpecArgs extends StringColumn(this) with PrimaryKey with ClusteringOrder with Ascending

  object json extends StringColumn(this)
}
```

**Comment by melodious** [comment]
@alexflav23 I made the code changes that you suggested but they did not change the generated CQL. I installed a local Cassandra database and tried to create the table with the CQL generated by Phantom:
```
cqlsh> CREATE KEYSPACE Excelsior
   ...   WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 3 };
cqlsh> CREATE TABLE Excelsior.FoobarTable (targetSite text, timePartition text, lastTimestamp timestamp, firstTimestamp timestamp, testType text, testSubtype text, name text, target text, testSpecArgs text, json text, PRIMARY KEY ((targetSite, timePartition), name, testSpecArgs, testSubtype, testType, lastTimestamp, target, firstTimestamp)) WITH CLUSTERING ORDER BY (lastTimestamp DESC, firstTimestamp DESC, testType ASC, testSubtype ASC, name ASC, target ASC, testSpecArgs ASC);
InvalidRequest: Error from server: code=2200 [Invalid query] message="The order of columns in the CLUSTERING ORDER directive must be the one of the clustering key (name must appear before lasttimestamp)"
```
I then re-arranged the PrimaryKey fields to match the clustering order and it created the table successfully:
```
cqlsh> CREATE TABLE Excelsior.MyFoobarTable (targetSite text, timePartition text, lastTimestamp timestamp, firstTimestamp timestamp, testType text, testSubtype text, name text, target text, testSpecArgs text, json text, PRIMARY KEY ((targetSite, timePartition), lastTimestamp, firstTimestamp, testType, testSubtype, name, target, testSpecArgs)) WITH CLUSTERING ORDER BY (lastTimestamp DESC, firstTimestamp DESC, testType ASC, testSubtype ASC, name ASC, target ASC, testSpecArgs ASC);
cqlsh> describe table excelsior.myfoobartable

CREATE TABLE excelsior.myfoobartable (
    targetsite text,
    timepartition text,
    lasttimestamp timestamp,
    firsttimestamp timestamp,
    testtype text,
    testsubtype text,
    name text,
    target text,
    testspecargs text,
    json text,
    PRIMARY KEY ((targetsite, timepartition), lasttimestamp, firsttimestamp, testtype, testsubtype, name, target, testspecargs)
) WITH CLUSTERING ORDER BY (lasttimestamp DESC, firsttimestamp DESC, testtype ASC, testsubtype ASC, name ASC, target ASC, testspecargs ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```

**Comment by alexflav23** [comment]
HI @melodious You are right, I've been looking in the wrong place, as per the PR title. The real issue is the primary key columns losing order, the clustering keys are fine. I'll add a ticket and look into that today, apologies.

**Comment by melodious** [comment]
@alexflav23 any update on when this might be complete? It's a blocker and I'll have to start coding some work-arounds if it's not going to be soon. Thanks!

**Comment by alexflav23** [comment]
@melodious Fixed in 2.2.0, going live today. https://github.com/outworkers/phantom/pull/630

**Comment by alexflav23** [comment]
@melodious It should hit Maven Central in 2 hours and Bintray in 20 minutes, you can add our Bintray resolver and remove it later in the day, but it would unblock you.

**Comment by melodious** [comment]
Thanks!

**Comment by alexflav23** [comment]
@melodious @rtatavarty Fixed in 2.2.0 already on Bintray, you  can use `Resolver.bintrayRepo("outworkers", "oss-releases")` for the next 2 hours while Maven Central artefact makes its way. It's them who take long, the build is finished on our side and everything is published.
